primitives/src/merkle_tree/field_based_mht/ramht/mod.rs

```diff
@@ -0,0 +1,75 @@
+use algebra::Field;
+use std::fmt::Debug;
+
+pub mod poseidon;
+
+//TODO: We can just keep this and remove the implementation at field_based_mht/mod.rs ?
+pub trait RandomAccessMerkleTree: Clone {
+    type Data: Field;
+    type MerklePath: Clone + Debug;
+
+    // Initialize this tree. The user must pass the maximum number of leaves the tree must
+    // support.
+    fn init(num_leaves: usize) -> Self;
+
+    // This function reset the pointing indices for the new element positions and processed positions
+    // Mainly used for benchmarking
+    fn reset(&mut self) -> &mut Self;
+
+    // This function appends a new leaf to the Merkle tree
+    fn append(&mut self, leaf: Self::Data) -> &mut Self;
+
+    // This function sets the leaf at `index` to `new_leaf`.
+    fn set(&mut self, index: usize, new_leaf: Self::Data) -> &mut Self;
+
+    // This function finalizes the computation of the Merkle tree and returns an updated
+    // copy of it. This method is idempotent, and calling it multiple times will
+    // give the same result. It's also possible to `update` with more inputs in between.
+    fn finalize(&self) -> Self;
+
+    // This function finalizes the computation of the Merkle tree
+    // Once this function is called, it is not possible to further update the tree.
+    fn finalize_in_place(&mut self) -> &mut Self;
+
+    // TODO: I propose to remove finalize_in_place and only keep finalize which will
+    //       return a "FinalizedMerkleTree" that will have the below functions. In
```

Comment:
I propose to remove finalize_in_place and only keep finalize which will
return a "FinalizedMerkleTree" that will have the below functions. In
this way we can keep the two things separated without having to use
a flag to check if we are in "edit" mode or in "read" mode.
The other solution would be to keep in this trait only the "edit" functions
and putting the "read" functions in the implementation, without exposing
the functions in an interface.
Otherwise we keep this mixed approach.

Reply:
I guess it has been resolved as I don't see the ramht directory anymore in the final commit.

Reply:
No because we still need to finalize the tree before being able to call a root or a merkle path on it, using the same instance.